Support `PartialType`

Right now if we do something like
```
[case testListInferred]
from typing import List
def f(n: int) -> List[int]:
    x = []
    for i in range(n):
        x.append(i)
    return x
```
we get a
`AssertionError: <class 'mypy.types.PartialType'> unsupported`

--- a/mypyc/genops.py
+++ b/mypyc/genops.py
@@ -31,7 +31,7 @@ from mypy.nodes import (
 )
 import mypy.nodes
 from mypy.types import (
-    Type, Instance, CallableType, NoneTyp, TupleType, UnionType, AnyType, TypeVarType,
+    Type, Instance, CallableType, NoneTyp, TupleType, UnionType, AnyType, TypeVarType, PartialType,
 )
 from mypy.visitor import NodeVisitor
 from mypy.subtypes import is_named_instance
@@ -233,6 +233,9 @@ class Mapper:
             # TODO: Erase to object if object has value restriction -- or union (once supported)?
             assert not typ.values, 'TypeVar with value restriction not supported'
             return self.type_to_rtype(typ.upper_bound)
+        elif isinstance(typ, PartialType):
+            assert typ.var.type is not None
+            return self.type_to_rtype(typ.var.type)
         assert False, '%s unsupported' % type(typ)
 
     def fdef_to_sig(self, fdef: FuncDef) -> FuncSignature:

--- a/test-data/genops-lists.test
+++ b/test-data/genops-lists.test
@@ -198,3 +198,24 @@ L3:
     goto L1
 L4:
     return l
+
+[case testPartialListType]
+def f() -> None:
+    x = []
+    x.append(1)
+[out]
+def f():
+    x, r0 :: list
+    r1 :: int
+    r2 :: object
+    r3 :: bool
+    r4, r5 :: None
+L0:
+    r0 = []
+    x = r0
+    r1 = 1
+    r2 = box(int, r1)
+    r3 = x.append(r2) :: list
+    r4 = None
+    r5 = None
+    return r5

--- a/test-data/genops-optional.test
+++ b/test-data/genops-optional.test
@@ -152,3 +152,41 @@ L1:
     return r4
 L2:
     return y
+
+[case testPartialOptionalType]
+def f(y: int) -> None:
+    x = None
+    if y == 1:
+        x = y
+    if x is not None:
+        y = x
+[out]
+def f(y):
+    y :: int
+    x :: optional[int]
+    r0 :: None
+    r1 :: int
+    r2 :: bool
+    r3 :: object
+    r4, r5 :: bool
+    r6 :: int
+    r7 :: None
+L0:
+    r0 = None
+    x = r0
+    r1 = 1
+    r2 = y == r1 :: int
+    if r2 goto L1 else goto L2 :: bool
+L1:
+    r3 = box(int, y)
+    x = r3
+L2:
+    r4 = x is None
+    r5 = !r4
+    if r5 goto L3 else goto L4 :: bool
+L3:
+    r6 = unbox(int, x)
+    y = r6
+L4:
+    r7 = None
+    return r7

